Feature/#6

mochang2 opened the pull request:
[이슈 링크](https://github.com/meopin-top/convey-your-mind-FE/issues/6)

주석으로 TODO 써 있는 부분은 기획이 덜 나왔거나 API 연동이 필요한 부분

## 내용
<컴포넌트>
* 메인 페이지(`src/app/page.tsx`)
* 로그인 없이 바로 롤링 페이퍼 쓰러 가기
  * `src/components/app/WithoutSignUp.tsx`
* 자체 로그인
  * `src/components/app/WithSignUp.tsx` 
  * `src/components/app/SignIn.tsx` 
  * `src/components/app/SignUp.tsx` 
  * `src/components/app/ConfirmedPopUp.tsx` <- 초기 로딩 속도 향상을 위해 lazy loading 적용
* OAuth 로그인
  * `src/components/app/WithOauth.tsx`

<공통 로직 추상화>
* 로그인 여부 확인 로직(임시 방편, `localStorage`에 `accessToken`이라는 key 값 존재 여부로 확인 예정 <- OAuth도 가능한지 확인이 필요한 사항)
  * `src/hooks/use-need-logged-in.ts`
* 인풋 state 관리 로직
  * `src/hooks/use-input.ts`
* 인풋 포커스 변경 로직
  * `src/hooks/use-focus.ts`
* 리액트 포털 사용 가능 여부 확인 로직
  * `src/hooks/use-portal.ts`

## 미구현 내용
([이슈 11](https://github.com/meopin-top/convey-your-mind-FE/issues/6)에서 보강할 예정)

* API 연동(`use-request.ts` 리팩토링)
* OAuth 토큰 관리

mochang2 committed:
remove: ci cd 확인용 페이지 제거

mochang2 committed:
remove: 홈 페이지 예시 코드 삭제

mochang2 committed:
rename: storage 예시 파일(index -> local-storage)

mochang2 committed:
remove: constants 예시 파일 삭제

mochang2 committed:
rename: scss 최상위 파일(app -> index)

mochang2 committed:
feat: 로그인 페이지에 필요한 asset(이미지) 추가

mochang2 committed:
docs: style 선언 순서 컨벤션 수정

mochang2 committed:
fix: scss 못 찾는 에러 해결

Commit:
fix: 기존 버전 도커 이미지 삭제하는 내용 추가

mochang2 committed:
feat: local storage key값 삭제하는 메서드 추가

mochang2 committed:
test: local storage 테스트 코드 추가

mochang2 committed:
feat: 로그인 페이지(메인 페이지) 구현

아래 내용은 아직 SB가 나오지 않아 미구현된 내용
- 로그인 시 유효성 검사
- 가입없이 공유코드로 바로 편지쓰기 화면에 구현되어야 할 버튼
- Oauth redirection 기능
- 슬로건
- 부제목

mochang2 committed:
test: 로그인 페이지(메인 페이지) 테스트 구현

mochang2 committed:
refactor: 반복되는 코드 use input으로 통일

mochang2 committed:
refactor: 공통적인 스타일 코드 변수, preset에 선언

mochang2 committed:
test: use-input 훅 테스트 코드 작성

mochang2 committed:
refactor: local storage 키값 타입 선언으로 제한

- 타입 선언(storage.d.ts)
- 메서드 타입 변경(local-storage.ts)
- 로그인 후 저장하는 키값 변경(WithSignUp.tsx)
- 테스트 변경(WithSignUp.test.tsx)
- 라우터 선언(route.ts)

mochang2 committed:
remove: 인증 관련 상수 파일 삭제(authentication.ts)

mochang2 committed:
feat: 로그인 확인 훅 구현(use-need-logged-in.ts)

mochang2 committed:
refactor: 라우터 이동 전 로그인 안내 얼럿 추가(use-need-logged-in.ts)

mochang2 committed:
test: 로그인 확인 훅 테스트 코드 작성(use-need-logged-in.test.ts)

mochang2 committed:
style: 툴팁 위치 반응형 대응(app.scss)

mochang2 committed:
refactor: 변수 선언 방식 컨벤션에 따라 변경(WithOAuth.tsx)

mochang2 committed:
refactor: portal 컴포넌트 분리

mochang2 committed:
test: portal 컴포넌트 테스트 코드 작성

mochang2 committed:
remove: portal loading 제거

mochang2 committed:
test: alert 컴포넌트 테스트 코드 작성(Alert.test.tsx)

mochang2 committed:
Merge feature/#7 into feature/#6

mochang2 committed:
refactor: 타입 import 변경(Alert.tsx)

Commit:
chore: 코드 에디터 설정 파일 형상 관리 제외(.gitignore)

Commit:
feat: float utility 추가(utility-first.scss)

Commit:
feat: 네이버 로고 추가(index.ts, naver-logo.webp)

mochang2 committed:
feat: 스타일 클래스 타입 선언(style.d.ts)

mochang2 committed:
feat: SecretInput 컴포넌트 생성

- 컴포넌트(SecretInput.tsx)
- barrel(index.ts)
- 스타일(index.scss, secret-input.scss)

mochang2 committed:
test: SecretInput 컴포넌트 테스트 코드 작성(SecretInput.test.tsx)

mochang2 committed:
test: 인풋 값 변경 추가(SecretInput.test.tsx)

mochang2 committed:
feat: SB 0.2 메인 페이지 구현(OAuth, 비밀번호 찾기, API 제외)

- icon asset(eye-close.svg, eye-open.svg, index.ts)
- 스타일 시트(index.scss, alert.scss, app.scss)
- 컴포넌트(ConfirmedPopUp.tsx, index.ts, SignIn.tsx, SignUp.tsx,
  WithOauth.tsx, WithoutSignUp.tsx, WithSignUp.tsx)
- 응답 코드 추가(response-code.ts)
- input에 포커스 조건 훅으로 선언(use-focus.ts)

mochang2 committed:
remove: 예시 fixture 삭제

mochang2 committed:
refactor: TProps export, lazy loading 문구 변경

- TProps export(Portal.tsx, ConfirmedPopUp.tsx)
- lazy loading named export(barrel) -> default export 로 변경.
  SecretInput은 lazy loading이 아니기 때문(SignUp.tsx)

mochang2 committed:
test: SB 0.2 메인 페이지 테스트 코드 작성

- 로그인 없이(WithoutSignUp.test.tsx)
- 로그인 있이(WithSignUp.test.tsx, SignIn.test.tsx, SignUp.test.tsx,
  ConfirmedPopUp.test.tsx)
- OAuth 이용해서(WithOauth.test.tsx)

mochang2 committed:
chore: OAuth env 추가(.env.example, push_ci-cd.yaml)

mochang2 committed:
feat: OAuth 로그인 추가

- 타입 선언(o-auth.d.ts)
- 스크립트 가져오기(page.tsx)
- 로그인 기능 추가(WithOauth.tsx)
- 스타일 시트(app.scss)
- 라우트 선언(route.ts)

mochang2 committed:
rename: o-auth.d.ts -> Oauth.d.ts

mochang2 committed:
test: naver, kakao 모킹(WithOauth.test.tsx)

mochang2 committed:
remove: 스타일 파일(color.d.ts)

mochang2 committed:
remove: 주석 제거(page.tsx)

JIH0LEE commented on src/components/app/SignUp.tsx:
```diff
@@ -0,0 +1,105 @@
+      /^[a-zA-Z0-9!@#$%^&*()-_=+{}\[\]|\\;:'",.<>/?]{6,20}$/
+    const passWordValidation =
+      /^(?=.*[a-zA-Z])(?=.*[0-9!@#$%^&*()-_=+{}\[\]|\\;:'",.<>/?])[a-zA-Z0-9!@#$%^&*()-_=+{}\[\]|\\;:'",.<>/?]{8,20}$/
+
```
userIdValidation, passwordValidation 상수들은 표기를 왜 카멜 케이스로 하셨는 지 궁금합니다.
대문자 스네이크 케이스를 사용하지 않은 이유도 알고 싶습니다.
EX) USER_ID_VALIDATION, PASSWORD_VALIDATION

JIH0LEE commented on src/components/app/SignIn.tsx:
```diff
@@ -0,0 +1,72 @@
+    if (code === SIGN_IN.SUCCESS) {
+      new Storage().set("accessToken", data.nickName) // TODO(remove): nickName이 있으면 로그인한 것
+    }
+
```
아마 현재 로그인을 성공하면 AccessToken을 쿠키에 넘겨줄 거에요

JIH0LEE commented on src/components/app/WithOauth.tsx:
```diff
@@ -0,0 +1,64 @@
+      })
+    }
+  }
+
```
해당 부분은 이번주 내로 리서치 끝내고 기능테스트 및 개발까지 완료해보겠습니다.

JIH0LEE commented on src/components/app/ConfirmedPopUp.tsx:
```diff
@@ -0,0 +1,136 @@
+      alert(message)
+    }
+  }
+
```
signUp() 로직이 SignUp.tsx가 아니라 ConfirmedPopUp.tsx에 있는 지 궁금합니다.

JIH0LEE commented on src/components/app/ConfirmedPopUp.tsx:
```diff
@@ -0,0 +1,136 @@
+    signUp()
+  }
+
+  async function signUp() {
+    // TODO: API 요청 보내면 loading 처리(button disabled), use-request 리팩토링
+    const {message, code} = await post("/users/sign-up", {
+      userId: userId.trim(),
+      password: password.trim(),
+      passwordCheck: confirmedPassword.trim(),
+    })
+
+    if (code === SIGN_UP.SUCCESS) {
+      // TODO: 정책 결정 필요
+    } else {
+      alert(message)
+    }
+  }
```
signUp() 로직이 SignUp.tsx가 아니라 ConfirmedPopUp.tsx에 있는 지 궁금합니다.

mochang2 replied:
회원가입 시 비밀번호 재확인이 팝업에서 이루어지기 때문입니다.

아마 위 내용보다 컴포넌트를 분리한 이유가 원하는 질문에 대한 대답이 될 것 같습니다.
제가 컴포넌트 나누는 기준은 크게 3가지 정도 됩니다.

1. 별도의 state를 관리하는 컴포넌트이다.
2. 컴포넌트를 분리하면서 비즈니스 로직과 관심사를 분리할 수 있다.
3. lazy loading이 가능한 요소이다.

`ConfirmedPopUp.tsx`는 위 3가지 요소를 다 만족해서 분리했습니다.
컴포넌트 네이밍에 대해서 궁금한 거라면 `SignIn.tsx`와 `SignUp.tsx`가 탭으로 구분되는 컴포넌트이기 때문에 해당 네이밍이 더 직관적으로 와닿을 거라고 생각했습니다.
만약 다른 네이밍이 더 낫다고 생각하시면 추천해주세요. 고려하겠습니다.

JIH0LEE replied:
만약 기획이 변경되어서 popup이 아닌 다른 컴포넌트에서 회원가입을 진행을 해야한다면 (ex. ConfirmedModal)  signUp 로직도 이동(변경)하는 건가요?  signUp이라고 하는 로직이 ConfirmedPopUp이라는 컴포넌트에 의존성이 생기는 것 같습니다.

mochang2 replied:
맞는 말씀입니다. 해당 로직 `SignUp.tsx`으로 이동 후 `signup` 함수를 props로 전달하는 방식으로 변경했습니다.

JIH0LEE commented on src/components/app/SignIn.tsx:
```diff
@@ -0,0 +1,72 @@
+    }
+  }
+
+  async function signIn() {
+    const {message, code, data} = await post("/users/sign-in", {
+      userId,
+      password,
+    })
+
+    if (code === SIGN_IN.SUCCESS) {
+      new Storage().set("accessToken", data.nickName) // TODO(remove): nickName이 있으면 로그인한 것
+    }
+
+    alert(message)
+  }
```
아마 현재 로그인을 성공하면 AccessToken을 쿠키에 넘겨줄 거에요

mochang2 replied:
넵~ 알겠습니다. 해당 내용은 #11 에서 수정하겠습니다.
프론트에서는 `localStorage`나 별도의 쿠키에서 로그인했는지 여부만 관리하겠습니다.
추후에 로그아웃에 관해서도 이야기 나누면 그 부분도 반영하도록 하겠습니다.

JIH0LEE commented on src/components/app/SignUp.tsx:
```diff
@@ -0,0 +1,105 @@
+    }
+  }
+
+  function checkValidation() {
+    const userIdValidation =
+      /^[a-zA-Z0-9!@#$%^&*()-_=+{}\[\]|\\;:'",.<>/?]{6,20}$/
+    const passWordValidation =
+      /^(?=.*[a-zA-Z])(?=.*[0-9!@#$%^&*()-_=+{}\[\]|\\;:'",.<>/?])[a-zA-Z0-9!@#$%^&*()-_=+{}\[\]|\\;:'",.<>/?]{8,20}$/
+
```
userIdValidation, passwordValidation 상수들은 표기를 왜 카멜 케이스로 하셨는 지 궁금합니다.
대문자 스네이크 케이스를 사용하지 않은 이유도 알고 싶습니다.
EX) USER_ID_VALIDATION, PASSWORD_VALIDATION

mochang2 replied:
음... JS에서는 `regExp.test('테스트하고 싶은 문자열')` 이런 식으로 정규 표현식을 활용합니다.
그래서 저는 자연스럽게 정규 표현식이 상수가 아니라 클래스 인스턴스나 함수의 일부분이라고 생각했습니다.
상수가 맞을까요...? ㅋㅎㅋ

JIH0LEE replied:
로직상 해당 변수의 값이 변경되지 않으면 상수라고 생각합니다. 이건 단순 궁금증으로 여쭤본거라서 괜찮습니다!

JIH0LEE commented on src/components/app/WithOauth.tsx:
```diff
@@ -0,0 +1,64 @@
+import {kakaoLogo, naverLogo} from "@/assets/images"
+import ROUTE from "@/constants/route"
+
+const WithOauth = () => {
+  const LOGO_SIZE = 72
+
+  // TODO: OAuth 토큰 관리 이야기 후 토큰 관리 추가하기
+  // redirect 이후 생기는 해시 값 또는 쿼리 스트링 관리할 방법 구상하기(네이버의 경우 이 내용이 토큰임)
+
+  useEffect(() => {
+    window.Kakao.init(process.env.NEXT_PUBLIC_KAKAO_JS_KEY)
+
+    new window.naver.LoginWithNaverId({
+      clientId: process.env.NEXT_PUBLIC_NAVER_CLIENT_ID,
+      callbackUrl: `${process.env.NEXT_PUBLIC_HOST}/${ROUTE.MY_PAGE}`,
+      isPopup: false,
+    }).init()
+  }, [])
+
+  function loginWithKakao() {
+    if (window.Kakao.isInitialized()) {
+      window.Kakao.Auth.authorize({
+        redirectUri: `${process.env.NEXT_PUBLIC_HOST}/${ROUTE.MY_PAGE}`,
+      })
+    }
+  }
+
```
해당 부분은 이번주 내로 리서치 끝내고 기능테스트 및 개발까지 완료해보겠습니다.

mochang2 replied:
넵~ 감사합니다. 저도 폰트 최적화 이슈 처리하고 시간되는 대로 프론트에서 작업해야 될 부분 확인하겠습니다.

mochang2 committed:
refactor: 회원가입 로직 이동

- 의존성 분리를 위해 props로 전달(SignUp.tsx -> ConfirmedPopUp.tsx)
- 테스트 코드 렌더링 수정(ConfirmedPopUp.test.tsx)

mochang2 committed:
Merge commit 'ef89bcb7a19669062a1de4af10ee5b15f22735d3' into feature/#6

mochang2 committed:
feat: fallback font 적용(app.scss)

j1mmyson commented on src/components/app/SignIn.tsx:
```diff
@@ -0,0 +1,72 @@
+      password,
+    })
+
+    if (code === SIGN_IN.SUCCESS) {
```
`code != SiGN_IN.SUCCESS` 일 경우 어떻게 처리되는지 궁금합니다~ 
아직 기획쪽에서 정해진게 없어서 비워두신걸까요??

mochang2 replied:
넵넵 기획쪽이 완성되면 채워나갈 예정입니다.
아마 `code === SIGN_IN.SUCCESS`면 로그인 여부를 로컬 스토리지나 쿠키에 담고 redirect 시키고 아니면,
`code !== SIGN_IN.SUCCESS`면 단순히 현재처럼 alert만 띄우거나 input을 초기화하는 등의 방식을 할 수 있지 않을까 싶네요.

j1mmyson commented on src/components/app/SignUp.tsx:
```diff
@@ -0,0 +1,120 @@
+  function handlePasswordInput(event: KeyboardEvent<HTMLInputElement>) {
+    const isEnterKeyDowned = event.key === "Enter"
+    if (isEnterKeyDowned) {
+      checkValidation()
```
유효성 검증을 하는 로직이 여러번 있을 것 같은데
validator 객체를 만들어서 여러 타입에 대한 메서드 구현하는 건 어떨까요??
@JIH0LEE 님 의견도 궁금합니다-!

mochang2 replied:
현재로서는 사용자 인풋을 받고 검증해야 되는 로직이 회원가입에밖에 없습니다.
그리고 그 유효성 검증이 총 3개이고, 각각의 유효성 검증에 대한 정규 표현식이 다릅니다.
그래서 재사용성이나 이런 쪽에 대해서는 생각지 않아 봤었습니다.

제안해주신 방법이 약간 아래와 같은 방법인가요?

```ts
interface Validator {
  validate: (target: string) => boolean;
}

class UserIdValidator implements Validator {
  static validation = /some regex/;

  validator(target: string) {
    return validation.test(target);
  }
}

class PasswordValidator implements Validator {
  static validation = /some regex/;

  validator(target: string) {
    return validation.test(target);
  }
}

class EmailValidator implements Validator {
  static validation = /some regex/;

  validator(target: string) {
    return validation.test(target);
  }
}
```

mochang2 replied:
아니면 아래와 같은 방법?

```ts
class Validator {
  #validation

  constructor(validation: RegExp) {
    this.#validation = validation;
  }

  validate(target: string) {
    return this.#validation.test(target);
  }
}
```


또는

```ts
// constants/user-input.ts
export const VALIDATOR = {
  USER_ID = /~~~/, // 정규 표현식
  PASSWORD = /~~~/,  // 정규 표현식
  EMAIL = /~~~/,  // 정규 표현식
}
```

j1mmyson replied:
오 넵 제가 생각했던건 첫번째 방법입니다.
근데 많이 쓰이지 않는다고 하면 굳이 말씀해주신 것 처럼 굳이 재사용 생각하지 않아도 될 것 같네요 👍

j1mmyson commented on src/hooks/use-need-logged-in.ts:
```diff
@@ -0,0 +1,18 @@
+import Storage from "@/store/local-storage"
+import ROUTE from "@/constants/route"
+
+export default function useNeedLoggedIn() {
```
이 함수는 어느 시점에 호출되는건지 궁금합니다-

mochang2 replied:
로그인 컴포넌트 구현한 김에 이러한 훅이 필요하겠다 싶어서 구현한 것으로, 로그인 여부를 확인해야 하는 페이지에서 공통적으로 필요한 로직을 처리하려고 선언한 훅입니다.
현재로서는 특정 조건을 확인해서 만족하지 않으면 로그인 페이지로 redirect 시키는 기능을 할 생각입니다.
지금은 완성되진 않았고 #11 에서 완성할 예정입니다.

mochang2 committed:
test: onSubmit props 적용(ConfirmedPopUp.test.tsx)

j1mmyson reviewed (approved):
LGTM 👍

mochang2 committed:
refactor: 정규 표현식 상수로 선언(input.ts, ConfirmedPopUp.tsx, SignUp.tsx)
